ignite/metrics/nlp/bleu.py

```diff
@@ -185,20 +200,78 @@ def _corpus_bleu(
         gm = bp * math.exp(math.fsum(s))
         return gm
 
+    def _sentence_bleu(self, references: Sequence[Sequence[Any]], candidates: Sequence[Any],) -> float:
+        return self._corpus_bleu([references], [candidates])
+
+    def _corpus_bleu(
+        self, references: Sequence[Sequence[Sequence[Any]]], candidates: Sequence[Sequence[Any]],
+    ) -> float:
+
+        p_numerators: Counter = Counter()
+        p_denominators: Counter = Counter()
+
+        hyp_length_sum, ref_length_sum = self._n_gram_counter(
+            references=references, candidates=candidates, p_numerators=p_numerators, p_denominators=p_denominators,
+        )
+
+        bleu_score = self._brevity_penalty_smoothing(
+            p_numerators=p_numerators,
+            p_denominators=p_denominators,
+            hyp_length_sum=hyp_length_sum,
+            ref_length_sum=ref_length_sum,
+        )
+
+        return bleu_score
+
     @reinit__is_reduced
     def reset(self) -> None:
-        self._sum_of_bleu = torch.tensor(0.0, dtype=torch.double, device=self._device)
-        self._num_sentences = 0
+
+        if self.average == "macro":
+            self._sum_of_bleu = torch.tensor(0.0, dtype=torch.double, device=self._device)
+            self._num_sentences = 0
+
+        if self.average == "micro":
+            self.p_numerators: Counter = Counter()
+            self.p_denominators: Counter = Counter()
+            self.hyp_length_sum = 0
+            self.ref_length_sum = 0
 
     @reinit__is_reduced
     def update(self, output: Tuple[Sequence[Sequence[Any]], Sequence[Sequence[Sequence[Any]]]]) -> None:
         y_pred, y = output
-        for _y_pred, _y in zip(y_pred, y):
-            self._sum_of_bleu += self._corpus_bleu(references=[_y], candidates=[_y_pred])
-            self._num_sentences += 1
+
+        if self.average == "macro":
+            for refs, hyp in zip(y, y_pred):
+                self._sum_of_bleu += self._sentence_bleu(references=refs, candidates=hyp)
+                self._num_sentences += 1
+
+        elif self.average == "micro":
+            hyp_lengths, ref_lengths = self._n_gram_counter(
+                references=y, candidates=y_pred, p_numerators=self.p_numerators, p_denominators=self.p_denominators
+            )
+            self.hyp_length_sum += hyp_lengths
+            self.ref_length_sum += ref_lengths
 
     @sync_all_reduce("_sum_of_bleu", "_num_sentences")
-    def compute(self) -> torch.Tensor:
-        if self._num_sentences == 0:
+    def _compute_macro(self) -> torch.Tensor:
+        if self._num_sentences == 0 and self.average == "macro":
```

Comment:
```suggestion
        if self._num_sentences == 0:
```
The test concerning `self.average` sounds useless.